tests/std/tests/LWG4105_ranges_ends_with_and_integer_class/test.cpp

```diff
@@ -0,0 +1,89 @@
+    static constexpr pair<int, int> short_haystack[] = {{4, 42}};
+    static constexpr pair<long, long> long_needle[]  = {{13, 1}, {13, 2}, {13, 4}};
+    static constexpr pair<long, long> short_needle[] = {{13, 2}, {13, 4}};
+    static constexpr pair<long, long> wrong_needle[] = {{13, 2}, {13, 3}};
+
+    template <test::signed_integer_like Diff1, testable_range In1, test::signed_integer_like Diff2, testable_range In2>
+    static constexpr void test_range_rediff() {
+        using ranges::ends_with, ranges::equal_to;
+
+        {
+            In1 r1{haystack};
+            In2 r2{long_needle};
+            const same_as<bool> auto match = ends_with(test::make_redifference_subrange<Diff1>(r1),
+                test::make_redifference_subrange<Diff2>(r2), equal_to{}, get_first, get_second);
+            assert(match);
+        }
+        {
+            In1 r1{haystack};
+            In2 r2{short_needle};
+            const same_as<bool> auto match = ends_with(test::make_redifference_subrange<Diff1>(r1),
+                test::make_redifference_subrange<Diff2>(r2), equal_to{}, get_first, get_second);
+            assert(match);
+        }
+        {
+            In1 r1{haystack};
+            In2 r2{wrong_needle};
+            const same_as<bool> auto match = ends_with(test::make_redifference_subrange<Diff1>(r1),
+                test::make_redifference_subrange<Diff2>(r2), equal_to{}, get_first, get_second);
+            assert(!match);
+        }
+        {
+            In1 r1{short_haystack};
+            In2 r2{short_needle};
+            const same_as<bool> auto match = ends_with(test::make_redifference_subrange<Diff1>(r1),
+                test::make_redifference_subrange<Diff2>(r2), equal_to{}, get_first, get_second);
+            assert(!match);
+        }
+        {
+            In1 r1{haystack};
+            In2 r2{span<pair<long, long>, 0>{}};
+            const same_as<bool> auto match = ends_with(test::make_redifference_subrange<Diff1>(r1),
+                test::make_redifference_subrange<Diff2>(r2), equal_to{}, get_first, get_second);
+            assert(match);
+        }
+    }
+
+    template <ranges::input_range In1, ranges::input_range In2>
+    static void call() {
+        if constexpr (testable_range<In1> && testable_range<In2>) {
+            using int_class = ranges::range_difference_t<ranges::iota_view<long long>>;
+
+            test_range_rediff<int_class, In1, short, In2>();
+            static_assert((test_range_rediff<int_class, In1, short, In2>(), true));
+
+            test_range_rediff<short, In1, int_class, In2>();
+            static_assert((test_range_rediff<short, In1, int_class, In2>(), true));
+        }
+    }
+};
+
+int main() {
+#if !defined(_PREFAST_) && !defined(__EDG__) // TRANSITION, GH-1030 and GH-3567
+    test_in_in<instantiator, const pair<int, int>, const pair<long, long>>();
+#endif // ^^^ no workaround ^^^

```

Comment:
This is *extremely* expensive, making it our second-most memory-hungry test, and I think we need to mitigate this:

```
D:\GitHub\STL\out\x64>cl /EHsc /nologo /W4 /std:c++latest /MTd /Od /bigobj /d1reportMemorySummary /I %STL%\tests\std\inc
lude %STL%\tests\std\tests\LWG4105_ranges_ends_with_and_integer_class\test.cpp
test.cpp
Persistent heap 1 size: 1065004200
Persistent heap 2 size: 16034992
Peak working set size: 3006615552
```

Can we follow the `#ifdef TEST_EVERYTHING` pattern and select a reasonably small set of "interesting" range combos for normal testing?

Reply:
Done. I'm now using the `#ifdef TEST_EVERYTHING` pattern.